### DISCUSSION TITLE
Criar sistema de Microfrontend (single-spa)

### DISCUSSION THREAD

**Discussion by jdgabriel** [frontendbr/forum#2360] [Perguntas e Respostas]
Fala devs, blza?

Estou estudando possibilidades no frontend, e cheguei até os microfrontends.

Tenho em minhas mãos 2 times de devs (2:2) e 6 Projetos (backend e frontend) para manter na empresa que estou, todos projetos internos (RH, Monitoramento, Estoque, e etc.).

Por várias vezes nos pegamos negando o deploy das aplicações por um time ter terminado a sprint e outro ainda não. Nesse sentido de deploys separados vejo vantagem de usar `MF`.

Ja usaram,  alguma dica? `Prós` e `Contras` de usar essa abordagem.
Não possuo muita experiência com frontend. Estou com medo kkk

**Comment by ViniciusmDias** [comment]
Então amigo, sempre tem aquele trade-offzinho...

Acredito que separar eles agora no começo não é algo muito vantajoso, visto que vais sair do problema de deploy e vai começar a entrar em outros problemas, como sub dominios, authenticação por cookie's (ou seja la o que vc for usar), custo, entre outros. Por isso, acredito que você pode começar criando o proximo frontend separando ele desse monolito que tu tem, acho que é um caminho mais interessante.

Pŕos: Melhor fluxo de deploy, poder usar mais tecnologias diferentes e testar algumas novas, melhor separação dos produtos.
Contras: Custo e Complexidade.

Provavelmente fazendo isso você começa abrir portas para varios outros tipos de iniciativa, como um design system, por exemplo, porém você tem mão de obra e maturidade o suficiente pra iniciar isso sem diminuir o número de entregas? Se sim, pode dale

**Comment by ghost** [comment]
Dúvida: sei que a pergunta foi pra microfront, mas precisaria de tudo isso? Claro que há ganhos com o desacoplamento das funcionalidades, mas cogitou sobre codarem sempre atrás de uma feature flag? Creio que seria uma solução rápida e de baixo custo.

Mas se quiser aproveitar a oportunidade para implementar microfront, diria pra ver bem que dependendo da solução, pode trazer problemas de performance devido ao baixo reaproveitamento de código.

Sugeriria cogitar de modularizar a aplicação, e cada módulo ser uma dependência do app, gerenciada por versionamento.

**Comment by jdgabriel** [comment]
Vejo que tenho que estudar muita coisa ainda, antes de tomar alguma decisão.
Vlw galera. 😄

**Reply by ghost** [reply]
Envolva e estude junto com os devs, independente da senioridade, pra ser uma solução que agrade a todos e todos saibam o pq da decisão de ter usado tal solução :3

Só em ti vc vai se sobrecarregar, e junto com eles podem surgir idéias boas que talvez sozinho nem tenha pensado. Eles também serão os que vão passar a arquitetura pra frente pra quando alguém novo vier fazer onboarding.

**Comment by fiuzagr** [comment]
@jdgabriel , acho que não é exatamente microfront que vai solucionar isso, até pq, pra chegar no MF, vc vai precisar modularizar o sistema, separar em domínios e desenhar os times para atuar com esses domínios. Feito isso, independente de ser MF ou não, vc vai conseguir os deploys continuos, pois cada time terá seu escopo isolado.

Feature flag, como @ghost propôs, ajuda muito. Mas, trará uma complexidade grande tbm, pois o código vai crescer e carregar muitas condicionais que tornarão ele mais complexo e dificil de manter. Tem até aqueles que dizem pra usar a feature flag enquanto desenvolve e depois tira tudo antes de fazer o deploy. Se o sistema em questão não está modular (é oq parece, devido aos times não atuarem em escopos/módulos isolados) então a feature flag vai ser bem dificil de manter.
